src/core/OIDNDenoiser.js

```diff
@@ -0,0 +1,282 @@
+
+	}
+
+	get weightsUrl() {
+
+		return this.denoiser.weightsUrl;
+
+	}
+
+	set weightsUrl( url ) {
+
+		this.denoiser.weightsUrl = url;
+
+	}
+
+	// Debugging
+	get denoiserDebugging() {
+
+		return this.denoiser.debugging;
+
+	}
+
+	set denoiserDebugging( v ) {
+
+		this.denoiser.debugging = v;
+
+	}
+
+	constructor( renderer ) {
+
+		this.renderer = renderer;
+		this.denoiser = new Denoiser( 'webgl', renderer.domElement );
+		this.denoiser.inputMode = 'webgl';
+		this.denoiser.outputMode = 'webgl';
+		this.denoiser.weightsUrl = 'https://cdn.jsdelivr.net/npm/denoiser/tzas';
+		this.denoiser.onProgress( progress => this.handleProgress( progress ) );
+		//this.denoiser.hdr = true;
+
+		this.percentage = 0;
+		this.isDenoising = false;
+		this.fadeTime = 500;
+		this.denoiserFinished = 0;
+		this.cleanAux = true;
+		this._useAux = true;
+		this.externalAux = false;
+		this.auxTextures = { albedo: null, normal: null };
+		this.progressListeners = new Set();
+
+		this.linearToSRGBMaterial = new MeshBasicMaterial( {
+			map: null,
+			transparent: true,
+			blending: NoBlending,
+		} );
+
+		// Material to blend between pathtracer and denoiser
+		this.blendToCanvasMaterial = new ClampedInterpolationMaterial( {
+			map: null,
+			transparent: true,
+			premultipliedAlpha: renderer.getContextAttributes().premultipliedAlpha,
+		} );
+
+		this.quad = new FullScreenQuad( this.linearToSRGBMaterial );
+		this.srgbPathTracedTarget = new WebGLRenderTarget( 1, 1, { colorSpace: SRGBColorSpace } );
+		this.outputTexture = null;
+
+	}
+
+	dispose() {
+
+		this.linearToSRGBMaterial.dispose();
+		this.blendToCanvasMaterial.dispose();
+		this.srgbPathTracedTarget.dispose();
+		if ( this.outputTexture ) this.outputTexture.dispose();
+		this.quad.dispose();
+
+	}
+
+	setAuxTextures( albedoTexture, normalTexture ) {
+
+		this.externalAux = true;
+		this.auxTextures.albedo = albedoTexture;
+		this.auxTextures.normal = normalTexture;
+
+	}
+
+	async denoise( rawPathTracedTexture, albedoTexture, normalTexture ) {
+
+		this.isDenoising = true;
+
+		// Adjust the height /width if changed from before
+		const { width, height } = rawPathTracedTexture.image;
+
+		this.denoiser.width = width;
+		this.denoiser.height = height;
+		this.srgbPathTracedTarget.setSize( width, height );
+
+		// set this on the object so we can get it later to split it
+		this.getLinearToSRGBTexture( rawPathTracedTexture, this.srgbPathTracedTarget );
+
+		// Extract the raw webGLTextures
+		const colorWebGLTexture = this.getWebGLTexture( this.srgbPathTracedTarget.texture );
+		const albedoWebGLTexture = this.getWebGLTexture( albedoTexture );
+		const normalWebGLTexture = this.getWebGLTexture( normalTexture );
+
+		// Run the denoiser
+		this.renderer.resetState();
+		const denoisedWebGLTexture = await this.denoiser.execute( colorWebGLTexture, albedoWebGLTexture, normalWebGLTexture );
+		this.renderer.resetState();
+
+		if ( ! this.outputTexture ) this.outputTexture = this.createOutputTexture();
```

Comment:
I'm not exactly sure what might be going on but when I move this intialization to the constructor I get the following error from `WebGLStateManager` in `denoiser` and the screen is white, instead. The width and height are changed to 1 in the function, as well.

I'm seeing the following warning logged after moving the function:

```
webglStateManager.ts:54 WebGL: INVALID_OPERATION: bindBuffer: attempt to use a deleted object
```

Reply:
It might be because in the constructor we create the Denoiser which normally has a "ready" callback that we currently ignore.
If you try to create that texture using the renderer you have tensorflow initializing the WebGL backend and the renderer trying to draw at the same time.

The two systems fight over the state and something breaks. 

That's just my guess from my earlier experience getting the two WebGL systems to play well.

If you don't want the conditional creation like it is now we can try adding the `onReady` listener to the denoiser creation before firing. I can play with this...

Reply:
Hm. Even if we move the creation of the texture to before the `Denoiser` creation it has the same issue.

I'm also not fully following how these systems could be fighting or in what way they're asynchronous. Is denoiser or OIDN using "requestAnimationFrame" or "setTimeout" or a webworker something? Otherwise no other JS in main thread can run until all those WebGL calls are issued.

Reply:
I say we just leave it where it is then

Reply:
We can do that for now - it's more about understanding what's going on and whether it's indicative of other issues. I'll just mark it with a comment and TODO.

Reply:
I think the normal texture color space, color blow out issues, and image corruption issue mentioned in https://github.com/gkjohnson/three-gpu-pathtracer/pull/663#pullrequestreview-2204761468 are things I'd like to address or understand more before merging. The rest we can make a new issue to track for later.

I'm a bit concerned about the image corruption because I think users will immediately run into this. We can control whether the path tracer renders or uses the context while denoising but not users and this is a bit of an odd restriction.